presto-main/src/main/java/com/facebook/presto/operator/aggregation/AccumulatorCompiler.java

```diff
@@ -235,6 +243,137 @@ private static void generateAddInput(
         body.ret();
     }
 
+    private static void generateAddInputWindowIndex(
+            ClassDefinition definition,
+            FieldDefinition stateField,
+            List<ParameterMetadata> parameterMetadatas,
+            MethodHandle inputFunction,
+            CallSiteBinder callSiteBinder)
+    {
+        // TODO: implement masking based on maskChannel field once Window Functions support DISTINCT arguments to the functions.
+
+        Parameter index = arg("index", WindowIndex.class);
+        Parameter channels = arg("channels", type(List.class, Integer.class));
+        Parameter startPosition = arg("startPosition", int.class);
+        Parameter endPosition = arg("endPosition", int.class);
+
+        MethodDefinition method = definition.declareMethod(a(PUBLIC), "addInput", type(void.class), ImmutableList.of(index, channels, startPosition, endPosition));
+        Scope scope = method.getScope();
+
+        Variable position = scope.declareVariable(int.class, "position");
+
+        Binding binding = callSiteBinder.bind(inputFunction);
+        BytecodeExpression invokeInputFunction = invokeDynamic(
+                BOOTSTRAP_METHOD,
+                ImmutableList.of(binding.getBindingId()),
+                "input",
+                binding.getType(),
+                getInvokeFunctionOnWindowIndexParameters(
+                        scope,
+                        inputFunction.type().parameterArray(),
+                        parameterMetadatas,
+                        stateField,
+                        index,
+                        channels,
+                        position));
+
+        method.getBody()
+                .append(new ForLoop()
+                        .initialize(position.set(startPosition))
+                        .condition(BytecodeExpressions.lessThanOrEqual(position, endPosition))
+                        .update(position.increment())
+                        .body(new IfStatement()
+                                .condition(anyParametersAreNull(scope, parameterMetadatas))
+                                .ifFalse(invokeInputFunction)))
+                .ret();
+    }
+
+    private static BytecodeExpression anyParametersAreNull(Scope scope,
+            List<ParameterMetadata> parameterMetadatas)
+    {
+        Variable index = scope.getVariable("index");
+        Variable channels = scope.getVariable("channels");
+        Variable position = scope.getVariable("position");
+
+        int inputChannel = 0;
+
+        BytecodeExpression isNull = constantFalse();
+        for (ParameterMetadata parameterMetadata : parameterMetadatas) {
+            BytecodeExpression getChannel = channels.invoke("get", Object.class, constantInt(inputChannel)).cast(int.class);
```

Comment:
Since this is only used in one block now, I'd just move it down to the just after `case INPUT_CHANNEL`